dataflow/src/domain/mod.rs

```diff
@@ -1915,7 +1933,18 @@ impl Domain {
                                     // that the hole for the key we're replaying ends up being
                                     // filled, even if that hole is empty!
                                     r.writer_mut().map(|wh| {
+                                        use core::data::SizeOf;
+
                                         for key in backfill_keys.iter() {
+                                            let size = wh.with_key(&key[..])
+                                                .try_find_and(|rs| {
+                                                    rs.iter().map(|r| r.deep_size_of()).sum()
+                                                })
+                                                .map(|r| r.0.unwrap_or(0))
+                                                .unwrap_or(0);
+                                            wh.mem_size =
+                                                wh.mem_size.checked_sub(size as usize).unwrap();
```

Comment:
`mark_filled` *should* only be called if the current state is empty, so I think this should (if anything) just assert that the state for that key is 0. Or am I missing something?

Reply:
`mark_filled` calls `self.handle.clear()`, so at least potentially throws away data. I can add an assertion there, but then I don't see what the purpose of `mark_filled` is at all -- it would have nothing to do?

Reply:
You can think of `mark_filled` as removing the thunk that indicates that there's a hole. What it *actually* does is insert `Vec::new` for the given key so that we know that it isn't missing even if the backfill produced no rows for that key. In all cases though, the current state for that key *should* be empty. Even if it does do `clear` though, can't you do the same thing as for eviction and sum up the size of the cleared rows?

Reply:
That's what I'm doing here. I cannot do it inside `mark_filled` due to lack of access to `self.mem_size`, and I have to do it as a lookup because `evmap::WriteHandle::clear` does not return anything as it just enqueues an op. We could change `evmap` to do an internal lookup and return the cleared rows and/or the size, I guess.

Reply:
Ah, i see what you mean. Yes, you would have to do a lookup before clearing (though `mark_filled` will have access to `mem_size` with the change below, so this code could move in there). That said, I think what I'd really do is simply assert that there are no rows for that key (like we do in local state: https://github.com/mit-pdos/distributary/blob/aebefadc1fa540c0a363a5c53cf418b17efeed8e/core/src/local/single_state.rs#L241).